app/scripts/controllers/transactions/IncomingTransactionHelper.ts

```diff
@@ -0,0 +1,271 @@
+    isEnabled?: () => boolean;
+    lastFetchedBlockNumbers?: Record<string, number>;
+    remoteTransactionSource: RemoteTransactionSource;
+    transactionLimit?: number;
+    updateTransactions?: boolean;
+  }) {
+    this.hub = new EventEmitter();
+
+    this.#blockTracker = blockTracker;
+    this.#getCurrentAccount = getCurrentAccount;
+    this.#getLocalTransactions = getLocalTransactions || (() => []);
+    this.#getNetworkState = getNetworkState;
+    this.#isEnabled = isEnabled ?? (() => true);
+    this.#isRunning = false;
+    this.#lastFetchedBlockNumbers = lastFetchedBlockNumbers ?? {};
+    this.#remoteTransactionSource = remoteTransactionSource;
+    this.#transactionLimit = transactionLimit;
+    this.#updateTransactions = updateTransactions ?? false;
+
+    // Using a property instead of a method to provide a listener reference
+    // with the correct scope that we can remove later if stopped.
+    this.#onLatestBlock = async (blockNumberHex: Hex) => {
+      try {
+        await this.#update(blockNumberHex);
+      } catch (error) {
+        log.error('Error while checking incoming transactions', error);
+      }
+    };
+  }
+
+  start() {
+    if (this.#isRunning) {
+      return;
+    }
+
+    if (!this.#canStart()) {
+      return;
+    }
+
+    this.#blockTracker.addListener('latest', this.#onLatestBlock);
+    this.#isRunning = true;
+  }
+
+  stop() {
+    this.#blockTracker.removeListener('latest', this.#onLatestBlock);
+    this.#isRunning = false;
+  }
+
+  async #update(latestBlockNumberHex?: Hex): Promise<void> {
+    if (!this.#canStart()) {
+      return;
+    }
+
+    const latestBlockNumber = parseInt(
+      latestBlockNumberHex || (await this.#blockTracker.getLatestBlock()),
+      16,
+    );
+
+    const fromBlock = this.#getFromBlock(latestBlockNumber);
+    const address = this.#getCurrentAccount();
+    const currentChainId = this.#getCurrentChainId();
+    const currentNetworkId = this.#getCurrentNetworkId();
+
+    let remoteTransactions = [];
+
+    try {
+      remoteTransactions =
+        await this.#remoteTransactionSource.fetchTransactions({
+          address,
+          currentChainId,
+          currentNetworkId,
+          fromBlock,
+          limit: this.#transactionLimit,
+        });
+    } catch (error: any) {
+      return;
+    }
+
+    if (!this.#updateTransactions) {
+      remoteTransactions = remoteTransactions.filter(
+        (tx) => tx.txParams.to?.toLowerCase() === address.toLowerCase(),
+      );
+    }
+
+    const localTransactions = this.#updateTransactions
+      ? this.#getLocalTransactions()
+      : [];
+
+    const newTransactions = this.#getNewTransactions(
+      remoteTransactions,
+      localTransactions,
+    );
+
+    const updatedTransactions = this.#getUpdatedTransactions(
+      remoteTransactions,
+      localTransactions,
+    );
+
+    if (newTransactions.length > 0 || updatedTransactions.length > 0) {
+      this.#sortTransactionsByTime(newTransactions);
```

Comment:
I know that this sort comes from the old code. But I couldn't get the idea of sorting txs since they aggregated into map on the listener, maybe you have an idea here?

Reply:
You're correct it's less useful in the extension given the map of transactions in state, but still good practice to provide the arrays in the event in a consistent order.